### PR TITLE
New management command - Manually move "submitted" report to dissemination

### PR DESCRIPTION
**Pull request by rnovak338**
Addresses part 2 of #4272 - providing an administrative command to move a SAC to the `disseminated` state.

The command is: `python manage.py move_to_disseminated --report_id ID_GOES_HERE`. This will look for a SAC with the `report_id` that you entered, and attempt to move it to the `disseminated` status - ONLY if it is stuck as `submitted` and passes cross-validation with no errors.

In the event that dissemination fails, or cross-validation fails, the SAC is returned to the `auditee_certified` state.

Additionally, this PR brings in some curation work from `jadudm/curation-api`. An app for `curation` has been created, which builts SQL scripts for tracking the before/after state of an audit. The above management command will track the audit's state before and after dissemination if it is successful.

## How to test

Before attempting the steps below, make sure you do a `docker-clean` and `docker-first-run`.

1. Create a SAC and bring it to the disseminated state.
2. Using a RDBMS, find this SAC, change `submission_status` to `submitted`, and ADD a row to `transition_name` and `transition_date`, which should be `disseminated` and a date for this action respectively.
3. Truncate the `dissemination_general` table to remove all dissemination data of this SAC. This will allow the management command to attempt to redisseminate without any issue.
4. Run the management command against your SAC.
5. Verify that the `fac_accepted_date` of the disseminated record matches the FIRST occurrence that the SAC was submitted (There should be two instances of `submitted` in `transition_name` and `transition_date` at this point).
6. Verify there is a record of this change in the `curation.record_version` table.

## PR Checklist: Submitter

- [ ]   Link to an issue if possible. If there’s no issue, describe what your branch does. Even if there is an issue, a brief description in the PR is still useful.
- [ ]   List any special steps reviewers have to follow to test the PR. For example, adding a local environment variable, creating a local test file, etc.
- [ ]   For extra credit, submit a screen recording like [this one](https://github.com/GSA-TTS/FAC/pull/1821).
- [ ]   Make sure you’ve merged `main` into your branch shortly before creating the PR. (You should also be merging `main` into your branch regularly during development.)
- [ ]   Make sure you’ve accounted for any migrations. When you’re about to create the PR, bring up the application locally and then run `git status | grep migrations`. If there are any results, you probably need to add them to the branch for the PR. Your PR should have only **one** new migration file for each of the component apps, except in rare circumstances; you may need to delete some and re-run `python manage.py makemigrations` to reduce the number to one. (Also, unless in exceptional circumstances, your PR should not delete any migration files.)
- [ ]   Make sure that whatever feature you’re adding has tests that cover the feature. This includes test coverage to make sure that the previous workflow still works, if applicable.
- [ ]   Make sure the full-submission.cy.js [Cypress test](https://github.com/GSA-TTS/FAC/blob/main/docs/testing.md#end-to-end-testing) passes, if applicable.
- [ ]   Do manual testing locally. Our tests are not good enough yet to allow us to skip this step. If that’s not applicable for some reason, check this box.
- [ ]   Verify that no Git surgery was necessary, or, if it was necessary at any point, repeat the testing after it’s finished.
- [ ]   Once a PR is merged, keep an eye on it until it’s deployed to dev, and do enough testing on dev to verify that it deployed successfully, the feature works as expected, and the happy path for the broad feature area (such as submission) still works.
- [ ]   Ensure that prior to merging, the working branch is up to date with main and the terraform plan is what you expect.

## PR Checklist: Reviewer

- [ ]   Pull the branch to your local environment and run `make docker-clean; make docker-first-run && docker compose up`; then run `docker compose exec web /bin/bash -c "python manage.py test"`
- [ ]   Manually test out the changes locally, or check this box to verify that it wasn’t applicable in this case.
- [ ]   Check that the PR has appropriate tests. Look out for changes in HTML/JS/JSON Schema logic that may need to be captured in Python tests even though the logic isn’t in Python.
- [ ]   Verify that no Git surgery is necessary at any point (such as during a merge party), or, if it was, repeat the testing after it’s finished.

The larger the PR, the stricter we should be about these points.

## Pre Merge Checklist: Merger

- [ ]   Ensure that prior to approving, the terraform plan is what we expect it to be. `-/+ resource "null_resource" "cors_header" ` should be destroying and recreating its self and ` ~ resource "cloudfoundry_app" "clamav_api" ` might be updating its `sha256` for the `fac-file-scanner` and `fac-av-${ENV}` by default.
- [ ]   Ensure that the branch is up to date with `main`.
- [ ]   Ensure that a terraform plan has been recently generated for the pull request.

### CODE REVIEW

**Comment by rnovak338** [comment]
@jadudm Any thoughts when it comes to testing or test coverage for this new command?

My approach felt very hacky - I added manual transitions to the FSM in-line to bring a report to the `submitted` state (skipping through each step without any data) to verify that cross-validation would fail. I also allowed a report from the `auditee_certified` state to go through (and bypassing the `submitted` state condition that is required by the command).

In other words, I manipulated the logic in the command in order to unit test, but I'm scratching my head on how to "recreate" the user experience for testing this command end to end.

File: backend/audit/management/commands/move_to_disseminated.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        # check for validation errors.
+        errors = sac.validate_full()
+        if errors:
+            logger.info(
+                f"Unable to disseminate report with validation errors: {report_id}."
+            )
+            logger.info(errors["errors"])
+
+            # return to auditee_certified.
+            sac_revert_from_submitted(sac)
```
Comment (jadudm) [inline backend/audit/management/commands/move_to_disseminated.py:67]: `enable_audit_curation()` around this, because it is a curatorial act.

File: backend/audit/management/commands/move_to_disseminated.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            disseminated = sac.disseminate()
+            # `disseminated` is None if there were no errors.
+            if disseminated is None:
+                sac_transition(None, sac, transition_to=STATUS.DISSEMINATED)
+        # END ATOMIC BLOCK
+        disable_audit_curation()
+
+        # IF THE DISSEMINATION SUCCEEDED
+        # `disseminated` is None if there were no errors.
+        if disseminated is None:
```
Comment (jadudm) [inline backend/audit/management/commands/move_to_disseminated.py:83]: Collapse to an `if/else`, because it is testing a boolean condition. Only one of these two can happen.

File: backend/audit/models/viewflow.py [hunk 3]
```diff
@@ -7,6 +7,25 @@
...
+    """
+    Transitions the submission_state for a SingleAuditChecklist back
+    to "auditee_certified" so the user can re-address issues and submit.
+    This should only be executed via management command.
+    """
+
+    if sac.submission_status == STATUS.SUBMITTED:
+        flow = SingleAuditChecklistFlow(sac)
+
+        flow.transition_to_auditee_certified()
```
Comment (jadudm) [inline backend/audit/models/viewflow.py:20]: Wrap:

```
with curation_tracking():
  sac.save()
```

Perhaps we should always wrap our `with` as close to the action as possible?

Comment (jadudm) [reply backend/audit/models/viewflow.py:20]: https://stackoverflow.com/questions/28157929/how-to-safely-handle-an-exception-inside-a-context-manager

We might want to do a bit more reading to make sure we don't accidentally leave tracking on.

We should probably `disable_audit_tracking` somwhere at startup, as an "always-just-in-case"...

File: backend/run.sh [hunk 4]
```diff
@@ -19,6 +20,11 @@ source tools/seed_cog_baseline.sh
 setup_env
 gonogo "setup_env"
 
+#####
```
Comment (jadudm) [inline backend/run.sh:23]: Move to the end/late in the boot sequence. After the whole API standup/migrate/teardown dance.

File: backend/tools/curation_audit_tracking_init.sh [hunk 5]
```diff
@@ -0,0 +1,9 @@
+source tools/util_startup.sh
+
+function curation_audit_tracking_init {
+    startup_log "CURATION_AUDIT_TRACKING" "BEGIN"
+    python manage.py curation_audit_tracking --init
```
Comment (jadudm) [inline backend/tools/curation_audit_tracking_init.sh:5]: Always disable audit tracking right after the init.

File: backend/run.sh (hunk 4 above)
Comment (jadudm) [reply backend/run.sh:23]: NEVERMIND. Leave this early in the sequence.

WHY?

What if we crashed, and the DB is in `enable` mode? We could enter the `migration` stage, run a big migration, and rewrite every row of `singleauditchecklist` That would all be documented as a curation, when it kinda isn't. (Eg. if we add a column with a new default value.) 

So, we should make sure it is disabled *before* the middle sequence, not after.

**Review by jadudm — DISMISSED** [review]
Excellent stuff.

We're going to add a `with` context manager for the enable/disable, because it might help prevent a condition where audit tracking is left on, either accidentally or because of a system error.

Other notes inline. Small changes. Looks great overall.
